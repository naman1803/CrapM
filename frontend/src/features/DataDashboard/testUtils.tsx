jest.mock('recharts', () => ({
  __esModule: true,
  ...jest.requireActual('recharts'),
  ResponsiveContainer: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>ResponsiveContainer</p>
      {children}
    </div>
  )),
}));
