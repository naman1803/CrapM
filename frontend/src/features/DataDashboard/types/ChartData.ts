// used for SeverityBarChart
export interface SeverityData {
    facilityOwner: string;
    severityLevel: string;
    severityScore: number;
}

// used for SeverityOverTimeChart
export interface SeverityOverTimeData extends SeverityData {
    date: string;
}

// one instance of ranking
export interface FacilityExceedanceData {
    facilityOwner: string;
    noOfExceedances: number;
}
