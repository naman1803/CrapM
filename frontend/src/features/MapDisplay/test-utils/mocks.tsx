import {
  LakeGeoPoint,
  MapElement,
  MapElementType,
  SewageGeoPoint,
} from '../types/MapElements';

export const mockLakeElements: MapElement[] = [
  {
    id: 1,
    dataType: 'lake',
    isSelectable: true,
    elementType: MapElementType.Marker,
    dataPoint: {
      coordinates: [0, 0],
      id: 1,
      name: 'Grrrr Lake',
      additional_notes: 'This is a lake',
    } as LakeGeoPoint,
  },
  {
    id: 2,
    dataType: 'lake',
    isSelectable: true,
    elementType: MapElementType.Marker,
    dataPoint: {
      coordinates: [0, 0],
      id: 2,
      name: 'Obama Lake',
    } as LakeGeoPoint,
  },
];

export const mockSewageElements: MapElement[] = [
  {
    id: 1,
    dataType: 'sewage',
    isSelectable: true,
    elementType: MapElementType.Circle,
    dataPoint: {
      coordinates: [0, 0],
      id: 1,
      facility_owner: 'Eric',
      radius: 100,
      colour: 'Blue',
    } as SewageGeoPoint,
  },
  {
    id: 2,
    dataType: 'sewage',
    isSelectable: true,
    elementType: MapElementType.Circle,
    dataPoint: {
      coordinates: [0, 0],
      id: 2,
      facility_owner: 'Myron',
      radius: 100,
      colour: 'Red',
    } as SewageGeoPoint,
  },
];
