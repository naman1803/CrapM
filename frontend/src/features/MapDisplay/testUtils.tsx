import { LeafletEventHandlerFnMap, LatLngExpression } from 'leaflet';

jest.mock('react-leaflet', () => ({
  __esModule: true,
  ...jest.requireActual('leaflet'),
  useMap: () => ({
    flyTo: jest.fn(),
    getZoom: jest.fn().mockReturnValue(13), // Add getZoom method returning default zoom level
    on: jest.fn(), // Mock on method for event handling
    off: jest.fn(), // Mock off method for removing event listeners
  }),
  MapContainer: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>Map Container</p>
      {children}
    </div>
  )),
  Marker: jest
    .fn()
    .mockImplementation(
      ({
        position,
        children,
        key,
        eventHandlers,
      }: {
        eventHandlers: LeafletEventHandlerFnMap;
        position: LatLngExpression;
        key: any;
        children: React.ReactNode;
      }) => (
        <div key={key}>
          <p>{`Marker-${position.toString()}`}</p>
          {Object.entries(eventHandlers).map(([keyE, value]) => (
            <button key={keyE} onClick={value}>
              {`button-${keyE}-${position.toString()}`}
            </button>
          ))}
          {children}
        </div>
      ),
    ),
  Popup: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>Popup</p>
      {children}
    </div>
  )),
  TileLayer: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>TileLayer</p>
      {children}
    </div>
  )),
}));
