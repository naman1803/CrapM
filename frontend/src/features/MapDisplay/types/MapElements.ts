import { LatLngExpression } from 'leaflet';

export enum MapElementType {
  Marker = 'marker',
  Circle = 'circle',
}

// Represent a point on the map
interface MapElementBase {
  id: number;
  dataType: string;
  isSelectable: boolean;
}

interface MarkerElement extends MapElementBase {
  elementType: MapElementType.Marker;
  dataPoint: GeoPoint;
}

interface CircleElement extends MapElementBase {
  elementType: MapElementType.Circle;
  dataPoint: GeoCircle;
}

export interface GeoPoint {
  coordinates: LatLngExpression;
}

interface GeoCircle extends GeoPoint {
  radius: number;
  colour: string;
}

export interface LakeGeoPoint extends GeoPoint {
  id: number;
  name: string;
  additional_notes?: string;
}

export interface SewageGeoPoint extends GeoCircle {
  id: number;
  facility_owner: string;
}

export type MapElement = MarkerElement | CircleElement;
