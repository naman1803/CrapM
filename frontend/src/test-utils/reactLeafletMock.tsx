import { LeafletEventHandlerFnMap, LatLngExpression } from 'leaflet';

jest.mock('react-leaflet', () => ({
  __esModule: true,
  ...jest.requireActual('leaflet'),
  useMap: () => ({ flyTo: jest.fn() }),
  MapContainer: jest.fn().mockImplementation(({ children }) => (
    <div data-testid='leaflet-map-container'>
      <p>Map Container</p>
      {children}
    </div>
  )),
  Circle: jest
    .fn()
    .mockImplementation(
      ({
        center,
        radius,
        eventHandlers,
        children,
        key,
      }: {
        center: LatLngExpression;
        radius: number;
        eventHandlers: LeafletEventHandlerFnMap;
        children: React.ReactNode;
        key: any;
      }) => (
        <div key={key} data-testid={`Circle-${center.toString()}`}>
          <p>{`Circle-${center.toString()}-radius-${radius}`}</p>
          {Object.entries(eventHandlers).map(([keyE, value]) => (
            <button
              key={keyE}
              onClick={value}
              data-testid={`circle-${center.toString()}-button-${keyE}`}
            >
              {`Circle-button-${keyE}-${center.toString()}`}
            </button>
          ))}
          {children}
        </div>
      ),
    ),
  Marker: jest
    .fn()
    .mockImplementation(
      ({
        position,
        children,
        key,
        eventHandlers,
      }: {
        eventHandlers: LeafletEventHandlerFnMap;
        position: LatLngExpression;
        key: any;
        children: React.ReactNode;
      }) => (
        <div key={key} data-testid={`Marker-${position.toString()}`}>
          <p>{`Marker-${position.toString()}`}</p>
          {Object.entries(eventHandlers).map(([keyE, value]) => (
            <button
              key={keyE}
              onClick={value}
              data-testid={`marker-${position.toString()}-button-${keyE}`}
            >
              {`button-${keyE}-${position.toString()}`}
            </button>
          ))}
          {children}
        </div>
      ),
    ),
  Popup: jest.fn().mockImplementation(({ children }) => (
    <div data-testid='leaflet-popup'>
      <p>Popup</p>
      {children}
    </div>
  )),
  TileLayer: jest.fn().mockImplementation(({ children }) => (
    <div data-testid='leaflet-tilelayer'>
      <p>TileLayer</p>
      {children}
    </div>
  )),
}));
