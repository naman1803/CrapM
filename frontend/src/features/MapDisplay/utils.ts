import { LakeData } from '@type/LakeData';
import {
  MapElement,
  LakeGeoPoint,
  SewageGeoPoint,
  MapElementType,
} from './types/MapElements';
import { SewageCoordData } from '@/types/SewageData';

export function processLakeMapElements(
  lakeMapElements: LakeData[],
): MapElement[] {
  const elements: MapElement[] = lakeMapElements.map(
    (lakeMapElement: LakeData) => {
      const lakePoint: LakeGeoPoint = {
        id: lakeMapElement.id,
        name: lakeMapElement.lakeName,
        coordinates: [lakeMapElement.latitude, lakeMapElement.longitude],
        ...(lakeMapElement?.additionalNotes
          ? { additional_notes: lakeMapElement.additionalNotes }
          : {}),
      };
      return {
        id: lakeMapElement.id,
        dataType: 'lake',
        isSelectable: true,
        elementType: MapElementType.Marker,
        dataPoint: lakePoint,
      };
    },
  );
  return elements;
}

export function processSewageMapElements(
  sewageMapElements: SewageCoordData[],
): MapElement[] {
  const elements: MapElement[] = sewageMapElements.map(
    (sewageMapElement: SewageCoordData) => {
      const sewagePoint: SewageGeoPoint = {
        id: sewageMapElement.id,
        facility_owner: sewageMapElement.facilityOwner,
        coordinates: [sewageMapElement.latitude, sewageMapElement.longitude],
        radius: sewageMapElement.radius,
        colour: sewageMapElement.colour,
      };
      return {
        id: sewageMapElement.id,
        dataType: 'sewage',
        isSelectable: true,
        elementType: MapElementType.Circle,
        dataPoint: sewagePoint,
      };
    },
  );
  return elements;
}
