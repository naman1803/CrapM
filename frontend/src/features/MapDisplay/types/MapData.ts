import { LatLngExpression } from 'leaflet';
import { MapElement } from './MapElements';

//used for maps
export interface MapData {
  points: MapElement[];
  selectedPoint: MapElement | null;
  center: LatLngExpression;
  title: string;
  mapStyles?: React.CSSProperties;
}
