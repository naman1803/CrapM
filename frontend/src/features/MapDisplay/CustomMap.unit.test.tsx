import './testUtils';

import { render } from '@testing-library/react';
import Map, { MapProps} from './CustomMap';
import '@/test-utils/reactLeafletMock';
import {
  LakeGeoPoint,
  MapElement,
  MapElementType,
  SewageGeoPoint,
} from './types/MapElements';
import { LatLngExpression } from 'leaflet';
import { processLakeMapElements, processSewageMapElements } from './utils';
import { SewageCoordData } from '@/types/SewageData';
import { LakeData } from '@/types/LakeData';


// jest.mock('react-leaflet', () => ({
//   __esModule: true,
//   ...jest.requireActual('leaflet'),
//   MapContainer: jest.fn().mockImplementation(({ children }) => (
//     <div>
//       <p>Map Container</p>
//       {children}
//     </div>
//   )),
//   Marker: jest.fn().mockImplementation(({ children, key }) => (
//     <div key={key}>
//       <p>Marker-{key}</p>
//       {children}
//     </div>
//   )),
//   Popup: jest.fn().mockImplementation(({ children }) => (
//     <div>
//       <p>Popup</p>
//       {children}
//     </div>
//   )),
//   TileLayer: jest.fn().mockImplementation(({ children }) => (
//     <div>
//       <p>TileLayer</p>
//       {children}
//     </div>
//   )),
// }));



const mockMapData: MapProps = {
  points: [
    {
      elementType: MapElementType.Marker,
      dataType: 'marker',
      id: 2,
      dataPoint: {
        coordinates: [0.0, 0.0] as LatLngExpression,
      },
      isSelectable: false,
    },
  ],
  selectedPoint: null,
  center: [43.53518, -80.21965],
  title: 'Test Map',
  onMapElementClick: jest.fn(),
  hoveredPoint: null,
};

describe('Map', () => {
  it('should render the map', () => {
    const { asFragment } = renderMap(mockMapData);
    expect(asFragment()).toMatchSnapshot();
  });
  
  it('should not render any markers when given no points', () => {
    const { queryAllByText } = renderMap({ ...mockMapData, points: [] });
    expect(queryAllByText(/Marker-/)).toHaveLength(0);
  });

});
//should be removed in next feature
describe('map utils', () => {
  it('should process sewage map elements correctly', () => {
    const sewageMapElements: SewageCoordData[] = [
      {
        id: 1,
        facilityOwner: 'Owner A',
        latitude: 34.0522,
        longitude: -118.2437,
        radius: 20,
        colour: 'Red',
      },
      {
        id: 2,
        facilityOwner: 'Owner B',
        latitude: 40.7128,
        longitude: -74.006,
        radius: 20,
        colour: 'Red',
      },
    ];

    const expectedOutput: MapElement[] = [
      {
        id: 1,
        dataType: 'sewage',
        isSelectable: true,
        elementType: MapElementType.Circle,
        dataPoint: {
          id: 1,
          facility_owner: 'Owner A',
          coordinates: [34.0522, -118.2437],
          radius: 20,
          colour: 'Red',
        } as SewageGeoPoint,
      },
      {
        id: 2,
        dataType: 'sewage',
        isSelectable: true,
        elementType: MapElementType.Circle,
        dataPoint: {
          id: 2,
          facility_owner: 'Owner B',
          coordinates: [40.7128, -74.006],
          radius: 20,
          colour: 'Red',
        } as SewageGeoPoint,
      },
    ];

    const result = processSewageMapElements(sewageMapElements);
    expect(result).toEqual(expectedOutput);
  });

  it('should return an empty array when input is empty', () => {
    const sewageMapElements: SewageCoordData[] = [];
    const expectedOutput: MapElement[] = [];
    const result = processSewageMapElements(sewageMapElements);
    expect(result).toEqual(expectedOutput);
  });
});

describe('processLakeMapElements', () => {
  it('should process lake map elements correctly', () => {
    const lakeMapElements: LakeData[] = [
      {
        id: 1,
        lakeName: 'Lake A',
        latitude: 34.05,
        longitude: -118.25,
        additionalNotes: 'Beautiful lake',
      },
      {
        id: 2,
        lakeName: 'Lake B',
        latitude: 36.16,
        longitude: -115.15,
      },
    ];

    const expectedOutput: MapElement[] = [
      {
        id: 1,
        dataType: 'lake',
        isSelectable: true,
        elementType: MapElementType.Marker,
        dataPoint: {
          id: 1,
          name: 'Lake A',
          coordinates: [34.05, -118.25],
          additional_notes: 'Beautiful lake',
        } as LakeGeoPoint,
      },
      {
        id: 2,
        dataType: 'lake',
        isSelectable: true,
        elementType: MapElementType.Marker,
        dataPoint: {
          id: 2,
          name: 'Lake B',
          coordinates: [36.16, -115.15],
        } as LakeGeoPoint,
      },
    ];

    const result = processLakeMapElements(lakeMapElements);
    expect(result).toEqual(expectedOutput);
  });

  it('should handle empty lake map elements array', () => {
    const lakeMapElements: LakeData[] = [];
    const expectedOutput: MapElement[] = [];
    const result = processLakeMapElements(lakeMapElements);
    expect(result).toEqual(expectedOutput);
  });
});

const renderMap = (props: MapProps) => render(<Map {...props} />);
