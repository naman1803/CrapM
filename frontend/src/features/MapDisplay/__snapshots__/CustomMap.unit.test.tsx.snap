// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map should render the map 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-j7qwjs"
  >
    <div>
      <p>
        Map Container
      </p>
      <div>
        <p>
          TileLayer
        </p>
      </div>
      <div>
        <p>
          Marker-0,0
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
